### PR TITLE
initial SystemDescriptionDocument

### LINKED ISSUE #70: Formalize Computation Verification Game using SecureETH SDD
# Bounty

Formalize Verifier & Enforcer and HydratedRuntime (depends on EVMRuntime)
*FIXME* also depends on Interceptor #54 

https://guidelines.secureth.org/project-planning/system-description
## Scope
- Documentation
- Brainstorming open details

## Gain for the project
Sanity 🙏 

## Roles
bounty gardener: @pinkiebell  / 5%
bounty worker: @pinkiebell  / 75%
bounty reviewer: @peara / 20%

### PR DESCRIPTION
**Pull request by pinkiebell**
WIP

Fixes #70

### CODE REVIEW

**Comment by peara** [comment]
@pinkiebell 
Hello :smile: 
I haven't coded for sometimes and now feel out of date.
Luckily you put up this documentation.
Do you might if I do some proofreading on this ?

**Comment by pinkiebell** [comment]
@peara Sure 😄

File: docs/SystemDescriptionDocument.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+The goal of this project is to verify off-chain EVM computation on-chain.
+In our particular case, it is used to correctly exit spending conditions from a side-chain to the main-chain
+and also a means of a PoS mechanism to secure the side-chain.
+
+Executing a program consists of many intermediate steps and in turn,
+the EVM consists of various elements which change their data after each single step (opcode) of the program.
+These intermediate states can be compared in an deterministic manner, by hashing the state of the EVM’s elements into a single state hash.
+This makes it possible to construct a merkle tree from all those state hashes, each one of them representing a single execution step.
+
+To find the first convergence of the execution (the first mismatch), we start a round-based path walking,
```
Comment (peara) [inline docs/SystemDescriptionDocument.md]: I think you mean "divergence"

File: docs/SystemDescriptionDocument.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# Overview
+The goal of this project is to verify off-chain EVM computation on-chain.
+In our particular case, it is used to correctly exit spending conditions from a side-chain to the main-chain
+and also a means of a PoS mechanism to secure the side-chain.
+
+Executing a program consists of many intermediate steps and in turn,
+the EVM consists of various elements which change their data after each single step (opcode) of the program.
+These intermediate states can be compared in an deterministic manner, by hashing the state of the EVM’s elements into a single state hash.
```
Comment (peara) [inline docs/SystemDescriptionDocument.md]: an -> a

File: docs/SystemDescriptionDocument.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+This makes it possible to construct a merkle tree from all those state hashes, each one of them representing a single execution step.
+
+To find the first convergence of the execution (the first mismatch), we start a round-based path walking,
+starting from the merkle root until we reach the leaves.
+
+For more information on how this works, take a look at [Merkelizer.md](./Merkelizer.md).
+
+## Example
+Let us assume a solver registers a execution result given the `code` to the program, any inputs (`callData`) and the merkle root hash of the execution.
```
Comment (peara) [inline docs/SystemDescriptionDocument.md]: a execution -> an execution
the code to the program -> the code of the program
I think 'any' can by dropped?

File: docs/SystemDescriptionDocument.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+A Verifier can now execute this program and compare if both merkle root hashes are the same.
+If the root hash are the same then both have the same result, if not, the Verifier can challenge the solver.
+
+
+# Actors
+
+## Solver
+The solver registers execution result with the `Enforcer` contract.
+This is a requirement if the solver wants to exit a spending condition to the main-chain.
```
Comment (peara) [inline docs/SystemDescriptionDocument.md]: I think it should be 
"exit some tokens under a spending condition"
isn't it?

File: docs/SystemDescriptionDocument.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+For every registered execution, the solver also needs to attach it with a bond.
+If the solver gets challenged and wins, solver owns the bond of the challenger.
+(Challenger's bond is slashed)
+
+If the solver's execution result does not get challenged in a given challenge period, the execution is assumed as valid.
+In the other case, if the solver does stop playing the verification game and the game is reaching the challenge period,
+the solver will automatically lose (`claimTimeout`).
+
+## Verifier
+The verifier is incentified for the very fact that proving a execution wrong, the verifier gets the bond of the solver.
```
Comment (peara) [inline docs/SystemDescriptionDocument.md]: I think it's worth noting that verifier could be the one being damaged if the solver committed something wrong?

Comment (pinkiebell) [reply docs/SystemDescriptionDocument.md]: 👍 Further explained below via new commit. Thanks for reviewing 😄

File: docs/SystemDescriptionDocument.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+(Solver's bond is slashed)
+
+If the challenger opens a dispute and stops playing the verification game or committed a wrong result,
```
Comment (pinkiebell) [inline docs/SystemDescriptionDocument.md]: @peara @johannbarbie 
Update is here. Is it correct that we will disburse the bonds always to the winner?

Comment (pinkiebell) [reply docs/SystemDescriptionDocument.md]: @peara Yes, we talked about that around the watercooler.
We have to keep track of verifierBond>disputeId>(maybe executionId)

Comment (pinkiebell) [reply docs/SystemDescriptionDocument.md]: Any improvement I could make here in this passage or are we good to go? 😃

Comment (peara) [reply docs/SystemDescriptionDocument.md]: I'm not sure. I'd like to have more details like usecases and who would be solver and verifiers. But I guess that might be too specific for this document?

Comment (pinkiebell) [reply docs/SystemDescriptionDocument.md]: +1 for more usecases

**Pull request merged**